### PR TITLE
Pre-call view and network test

### PR DESCRIPTION
**Pull request by jeffswartz**

### CODE REVIEW

File: web/js/precallView.js [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+          break;
+        case 'initialVideoSwitch':
+          if (!initialVideoSwitch.classList.contains('activated')) {
+            setSwitchStatus(true, true, 'Video', 'roomView:initialVideoSwitch');
+          } else {
+            setSwitchStatus(false, true, 'Video', 'roomView:initialVideoSwitch');
+          }
+          break;
+      }
+    });
```
Comment (maikthomas) [inline web/js/precallView.js]: Again here just a suggestion, you could do something along these lines.
```

 publishSettings.addEventListener('click', function (e) {
  var switches = {
    initialVideoSwitch: {
      name: 'Video',
      element: document.querySelector('#initialVideoSwitch'),
      event: 'roomView:initialVideoSwitch'
    },
    initialAudioSwitch: {
      name: 'Audio',
      element:  document.querySelector('#initialAudioSwitch'),
      event: 'roomView:initialAudioSwitch'
    }
  }
  var elem = e.target;
  elem.blur();
  // pointer-events is not working on IE so we can receive as target a child
  elem = HTMLElems.getAncestorByTagName(elem, 'a');
  if (elem) {
    var switch_ = switches[elem.id];
    var activated = selectedSwitch.element.classList.contains('activated');
    setSwitchStatus(!activated, true, switch_.name, switch_.event);
  }
```

File: web/js/precallView.js [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+
+    var videoPreviewElement = document.getElementById('video-preview');
+    var videoPreviewNameElement = document.getElementById('video-preview-name');
+    videoPreviewElement.addEventListener('mouseover', function () {
+      videoPreviewNameElement.style.opacity = 1;
```
Comment (maikthomas) [inline web/js/precallView.js]: I think this would be better in the css using `:hover`

File: web/js/precallView.js [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+    });
+  }
+
+  var eventHandlers = {
+    'PrecallController:endPrecall': function () {
+      _model.addEventListener('value', render);
+      render();
+    },
+    'PrecallController:audioOnly': function () {
+      setSwitchStatus(false, true, 'Video', 'roomView:initialVideoSwitch');
```
Comment (maikthomas) [inline web/js/precallView.js]: Is there a way we could make this function clearer? I'm not a fan of listing booleans and having to get check what each one does.

File: web/js/precallView.js [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+    document.getElementById('pre-call-test-results').style.display = 'block';
+    document.getElementById('audio-bitrate').innerText =
+      Math.round(results.audio.bitsPerSecond / 1000);
+    if (results.video && results.text !== 'Your bandwidth can support audio only.') {
+      document.getElementById('video-bitrate').innerText =
+        Math.round(results.video.bitsPerSecond / 1000);
+      packetLossStr = isNaN(results.video.packetLossRatio) ? '' :
+        Math.round(100 * results.video.packetLossRatio) + '% packet loss';
+      document.getElementById('precall-video-packet-loss').innerText = packetLossStr;
+      audioResultsElem.style.width = '50%';
```
Comment (maikthomas) [inline web/js/precallView.js]: Again wouldn't this be better off in the .less?

File: web/js/precallView.js (hunk 2 above)
Comment (jeffswartz) [reply web/js/precallView.js]: @maikthomas, The`mouseover` target (`videoPreviewElement`) is different than the element for which we are setting the opacity (`videoPreviewNameElement`). So, CSS `:hover` won't work.

File: web/js/precallView.js [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+      var elem = e.target;
+      switch (elem.id) {
+        case 'precall-close':
+          preCallTestResults.style.display = 'none';
+          break;
+        case 'retest':
+          preCallTestResults.style.display = 'none';
+          Utils.sendEvent('roomView:retest');
+          break;
+      }
```
[the remark this answers is not in the document]
Comment (jeffswartz) [reply web/js/precallView.js]: Nope. This change would set `style.display` to `'none'` if any part of the pre-call test results is clicked (which isn't what we want).

File: web/js/precallView.js (hunk 1 above)
Comment (jeffswartz) [reply web/js/precallView.js]: Yes, but if there were other `a` tags in the publish-settings element, this code could result a "Cannot read property 'element' of undefined" error.

File: web/js/precallView.js (hunk 3 above)
Comment (jeffswartz) [reply web/js/precallView.js]: I've decided to pull the second Boolean parameter. The event can always bubble up.

File: web/js/precallView.js (hunk 4 above)
Comment (jeffswartz) [reply web/js/precallView.js]: Actually, we don't need this CSS adjustment anymore. We always display both the Audio and Video results (with video saying "no video" if there is no video).

File: web/js/precallView.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (maikthomas) [reply web/js/precallView.js]: Ah ok, there's more values of `elem.id` than those in the switch, my bad!

File: web/js/precallView.js (hunk 2 above)
Comment (maikthomas) [reply web/js/precallView.js]: Hmm ok, so the alternative would be to set a data-attr or add a class but we would still need this code so not worth much. Thanks for clarifying.

File: web/js/precallView.js (hunk 4 above)
Comment (maikthomas) [reply web/js/precallView.js]: Awesome 👍

File: web/js/precallView.js (hunk 3 above)
Comment (maikthomas) [reply web/js/precallView.js]: Perfect 👍

File: web/js/precallView.js (hunk 1 above)
Comment (maikthomas) [reply web/js/precallView.js]: I see, same as for my other comment, more possible outcomes than the switch options 👍

Comment (maikthomas) [reply web/js/precallView.js]: That said, I'll still support gathering variables and ending with a single call to the function.

**Pull request merged**